File: src/presentation/dto/model/ProductDetail.dto.ts
import { ApiProperty } from "@nestjs/swagger";
import { Type } from "class-transformer";
import { IsBoolean, IsNumber, IsString, Matches, Min, ValidateNested } from "class-validator";

export class DimensionsDTO {
  @ApiProperty({ description: "가로 길이", example: 200 })
  @IsNumber()
  @Min(0)
  width: number;

  @ApiProperty({ description: "세로 길이", example: 85 })
  @IsNumber()
  @Min(0)
  height: number;

  @ApiProperty({ description: "깊이", example: 90 })
  @IsNumber()
  @Min(0)
  depth: number;
}

export class AdditionalInfoDTO {
  @ApiProperty({ description: "조립 필요 여부", example: true })
  @IsBoolean()
  assembly_required: boolean;

  @ApiProperty({ description: "조립 시간", example: "30분" })
  @Matches(/^(\d+시간)?(\d+분)?(\d+초)?$/, {
    message: "조립 시간 형식은 'X시간 Y분 Z초'이어야 합니다.",
  })
  assembly_time: string;
}

export default class ProductDetailDTO {
  @ApiProperty({ description: "무게", example: 25.5 })
  @IsNumber()
  @Min(0)
  weight: number;

  @ApiProperty({ description: "크기 정보", type: DimensionsDTO })
  @ValidateNested()
  @Type(() => DimensionsDTO)
  dimensions: DimensionsDTO;

  @ApiProperty({ description: "재료", example: "가죽, 목재, 폼" })
  @IsString()
  materials: string;

  @ApiProperty({ description: "원산지", example: "대한민국" })
  @IsString()
  country_of_origin: string;

  @ApiProperty({ description: "보증 정보", example: "2년 품질 보증" })
  @IsString()
  warranty_info: string;

  @ApiProperty({ description: "관리 지침", example: "마른 천으로 표면을 닦아주세요" })
  @IsString()
  care_instructions: string;

  @ApiProperty({ description: "추가 정보", type: AdditionalInfoDTO })
  @ValidateNested()
  @Type(() => AdditionalInfoDTO)
  additional_info: AdditionalInfoDTO;
}

File: wantedcqrs/src/main/java/com/pawn/wantedcqrs/product/dto/ProductDetailDto.java
package com.pawn.wantedcqrs.product.dto;

import com.pawn.wantedcqrs.product.entity.Product;
import com.pawn.wantedcqrs.product.entity.ProductDetail;
import lombok.Data;

import java.math.BigDecimal;

@Data
public class ProductDetailDto {

    private Long id;

    private Long productId;

    private BigDecimal weight;

    private ProductDetailDto.Dimensions dimensions;

    private String materials;

    private String countryOfOrigin;

    private String warrantyInfo;

    private String careInstructions;

    private ProductDetailDto.AdditionalInfo additionalInfo;

    @Data
    public static class Dimensions {

        private int width;

        private int height;

        private int depth;

        public Dimensions(int width, int height, int depth) {
            this.width = width;
            this.height = height;
            this.depth = depth;
        }

    }

    @Data
    public static class AdditionalInfo {

        private boolean assemblyRequired;

        private String assemblyTime;

        public AdditionalInfo(boolean assemblyRequired, String assemblyTime) {
            this.assemblyRequired = assemblyRequired;
            this.assemblyTime = assemblyTime;
        }

    }

    public ProductDetail toEntity() {
        return ProductDetail.builder()
                .id(id)
//                .product(product)
                .weight(weight)
                .dimensions(ProductDetail.Dimensions.builder()
                        .depth(dimensions.depth)
                        .width(dimensions.width)
                        .height(dimensions.height)
                        .build()
                ).materials(materials)
                .countryOfOrigin(countryOfOrigin)
                .warrantyInfo(warrantyInfo)
                .careInstructions(careInstructions)
                .additionalInfo(ProductDetail.AdditionalInfo.builder()
                        .assemblyRequired(additionalInfo.assemblyRequired)
                        .assemblyTime(additionalInfo.assemblyTime)
                        .build())
                .build();
    }

}
